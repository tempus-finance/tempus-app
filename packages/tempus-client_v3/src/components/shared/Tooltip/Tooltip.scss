@use "sass:math";

$arrowWidth: 16px;

.tc__tooltip {
  min-width: 100%;
  position: absolute;
  border-radius: 4px;
  box-shadow: var(--downWeakShadow);
  border: 1px solid var(--tooltipBorder);
  background: var(--tooltipBackground);

  // the CSS variable are here for React components to override
  --anchorWidth: 16px;

  &::before1 {
    content: '';
    display: block;
    position: absolute;
    width: $arrowWidth;
    height: $arrowWidth;
    border: 1px solid var(--tooltipBorder);
    background: var(--tooltipBackground);
    transform: translateY(calc($arrowWidth / 2)) rotate(45deg);
    transform-origin: center;
    border-bottom: 0;
    border-right: 0;
  }

  &::before {
    content: '';
    position: absolute;
    width: $arrowWidth;
    height: $arrowWidth;
    background: var(--tooltipBorder);
    -webkit-clip-path: polygon(50% 30%, 0% 100%, 100% 100%);
    clip-path: polygon(50% 30%, 0% 100%, 100% 100%);
  }

  &::after {
    content: '';
    position: absolute;
    width: calc($arrowWidth - 2px);
    height: calc($arrowWidth - 2px);
    background: var(--tooltipBackground);
    bottom: 100%;
    -webkit-clip-path: polygon(50% 30%, 0% 100%, 100% 100%);
    clip-path: polygon(50% 30%, 0% 100%, 100% 100%);
  }

  @each $horizontal-orientation in ('left', 'right') {
    $opposite-horizontal-orientation: if($horizontal-orientation == 'left', 'right', 'left');

    &.tc__tooltip__bottom-#{$horizontal-orientation} {
      transform: translateY(calc(100% + $arrowWidth));
      bottom: 0px;
      #{$opposite-horizontal-orientation}: 0px;

      &::before {
        #{$horizontal-orientation}: calc(100% - $arrowWidth/2 - max(var(--anchorWidth) / 2, $arrowWidth/2));
        bottom: 100%;
      }

      &::after {
        #{$horizontal-orientation}: calc(100% - $arrowWidth/2 - max(var(--anchorWidth) / 2, $arrowWidth/2) + 1px);
        bottom: 100%;
      }
    }
  }

  &.tc__tooltip__bottom-center {
    display: flex;
    justify-content: center;
    transform: translate(-50%, calc(100% + $arrowWidth));
    left: 50%;
    bottom: 0;

    &::before {
      bottom: 100%;
    }

    &::after {
      bottom: 100%;
    }
  }
}
