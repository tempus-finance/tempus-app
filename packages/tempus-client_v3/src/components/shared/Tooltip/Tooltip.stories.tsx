import { useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Tooltip from './Tooltip';

export default {
  title: 'Tooltip',
  component: Tooltip,
  argTypes: {
    placement: {
      control: {
        type: 'select',
        options: ['bottom-left', 'bottom-right'],
      },
    },
  },
} as ComponentMeta<typeof Tooltip>;

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '50px',
};

const tooltipContentStyle = {
  width: '300px',
  padding: '16px',
};

const Template: ComponentStory<typeof Tooltip> = props => {
  const [tooltipOpen, setTooltipOpen] = useState(false);

  const toggleTooltip = () => {
    setTooltipOpen(prevState => !prevState);
  };

  return (
    <div style={style}>
      <button type="button" onClick={toggleTooltip} style={{ position: 'relative' }}>
        Click me!
        {/* Tooltip should be placed inside it's anchor element - and anchor element must have relative position */}
        <Tooltip {...props} open={tooltipOpen}>
          {/* Content of the tooltip is placed inside tooltip as it's child element */}
          <div style={tooltipContentStyle}>Tooltip content</div>
        </Tooltip>
      </button>
    </div>
  );
};

export const Primary = Template.bind({});
Primary.args = {
  placement: 'bottom-right',
};
