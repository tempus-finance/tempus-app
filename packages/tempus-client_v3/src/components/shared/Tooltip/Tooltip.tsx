import { FC } from 'react';

import './Tooltip.scss';

export interface TooltipProps {
  open: boolean;
}

const Tooltip: FC<TooltipProps> = props => {
  const { open, children } = props;

  if (!open) {
    return null;
  }

  return <div className="tc__tooltip">{children}</div>;
};
export default Tooltip;
