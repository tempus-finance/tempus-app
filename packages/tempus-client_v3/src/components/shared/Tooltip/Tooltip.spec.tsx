import { render } from '@testing-library/react';
import Tooltip, { TooltipProps } from './Tooltip';

const defaultProps: TooltipProps = {
  open: true,
};

const tooltipContentText = 'Tooltip Content';

const subject = (props: TooltipProps) =>
  render(
    <Tooltip {...props}>
      <p>{tooltipContentText}</p>
    </Tooltip>,
  );

describe('Tooltip', () => {
  it('renders child elements as content when tooltip is open', () => {
    const { container, getByText } = subject(defaultProps);

    const tooltipContent = getByText(tooltipContentText);

    expect(tooltipContent).not.toBeNull();

    expect(container).toMatchSnapshot();
  });

  it('does not render child elements as content when tooltip is closed', () => {
    const { container, queryByText } = subject({ open: false });

    const tooltipContent = queryByText(tooltipContentText);

    expect(tooltipContent).toBeNull();

    expect(container).toMatchSnapshot();
  });
});
