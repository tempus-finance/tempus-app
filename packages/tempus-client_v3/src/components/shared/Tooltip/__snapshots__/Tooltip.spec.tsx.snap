// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip does not render child elements as content when tooltip is closed 1`] = `<div />`;

exports[`Tooltip renders child elements as content when tooltip is open 1`] = `
<div>
  <div
    class="tc__tooltip tc__tooltip__bottom-right"
  >
    <p>
      Tooltip Content
    </p>
  </div>
</div>
`;
